src/tribler-core/tribler_core/modules/metadata_store/community/eva_protocol.py

```diff
@@ -0,0 +1,502 @@
+    EVA supports timeouts as well.
+
+    The data size limit can be calculated as "block_size * 4294967295" where
+    4294967295 is the max segment number (4B unsigned int).
+    """
+
+    def eva_init(  # pylint: disable=too-many-arguments
+        self,
+        block_size=1000,  # please keep in mind that ipv8 adds approx. 177 byte to each packet
+        window_size_in_blocks=16,
+        start_message_id=186,
+        retransmit_interval_in_sec=3,
+        retransmit_attempt_count=3,
+        scheduled_send_interval_in_sec=5,
+        timeout_interval_in_sec=10,
+        binary_size_limit=1024 * 1024 * 1024,
+    ):
+
+        """Init should be called manually within his parent class.
+
+        Args:
+            block_size: a single block size
+            window_size_in_blocks: size of consecutive blocks to send
+            start_message_id: a started id that will be used to assigning
+                protocol's messages ids
+            retransmit_interval_in_sec: an interval until the next attempt
+                to retransmit a message in case its delivery is not confirmed
+            retransmit_attempt_count: limit for retransmit attempts
+            scheduled_send_interval_in_sec: an interval until the next attempt
+                to sent data in case it was scheduled before via "eva_send" method
+            timeout_interval_in_sec: an interval after which
+                the transfer is considered "dead" and will be terminated
+            binary_size_limit: limit for binary data size. In case of this limit
+                will be exceeded, the exception will be returned via a registered
+                error handler
+        """
+        # define instance's variables
+        self.eva_store = SimpleNamespace(
+            send_complete_callbacks=set(),
+            receive_callbacks=set(),
+            error_callbacks=set(),
+            incoming=dict(),
+            outgoing=dict(),
+            scheduled=defaultdict(deque),
+            block_size=block_size,
+            window_size=window_size_in_blocks,
+            retransmit_interval_in_sec=retransmit_interval_in_sec,
+            retransmit_attempt_count=retransmit_attempt_count,
+            timeout_interval_in_sec=timeout_interval_in_sec,
+            scheduled_send_interval_in_sec=scheduled_send_interval_in_sec,
+            binary_size_limit=binary_size_limit,
+            retransmit_enabled=True,
+            terminate_by_timeout_enabled=True,
+            nonce=0,
+        )
+
+        # specify messages' ids
+        _assign_id_to_messages(start_message_id, [WriteRequest, Acknowledgement, Data, Error])
+
+        # add messages' handlers
+        self.add_message_handler(WriteRequest, self.on_eva_write_request)
+        self.add_message_handler(Acknowledgement, self.on_eva_acknowledgement)
+        self.add_message_handler(Data, self.on_eva_data)
+        self.add_message_handler(Error, self.on_eva_error)
```

Comment:
Quick note: you don't need to change the `msg_id` of the class instances (which could screw with other community instances doing the same). You can specify the handlers by message number instead:

```suggestion
        # add messages' handlers
        self.add_message_handler(start_message_id, self.on_eva_write_request)
        self.add_message_handler(start_message_id + 1, self.on_eva_acknowledgement)
        self.add_message_handler(start_message_id + 2, self.on_eva_data)
        self.add_message_handler(start_message_id + 3, self.on_eva_error)
```

Reply:
My bad, forgot to mention that you're not allowed to use `self.ez_send()` anymore then.

```python
self.ez_send(peer, Acknowledgement(0, transfer.nonce))
```
⬇️ (with `self.ack_msg_id = start_message_id + 1`)
```python
self.endpoint.send(peer.address, self.ezr_pack(self.ack_msg_id, Acknowledgement(0, transfer.nonce)))
```

Reply:
I see. 
I'll rewrite all the occurrences because probably it seems like a single correct way to implement a dynamic message-id assignation (in case this mixin can be used by multiple classes in the future).